src/lib/Scenes/Home/Components/ForYou/index.tsx

```diff
@@ -15,80 +15,124 @@ interface DataSourceRow {
+  errors?: string // TODO: Wire up to UI handler
   isRefreshing: boolean
-  dataSource: ListViewDataSource
+}
+
+interface Module {
+  props: {
+    rail: any
+    relay: RelayProps
+  }
+  state: any
 }
 
 export class ForYou extends React.Component<Props, State> {
   listView?: ListView | any
   currentScrollOffset?: number = 0
+  modules: Module[] = []
 
-  constructor(props) {
-    super(props)
-
-    const dataSource: ListViewDataSource = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 })
-
-    this.state = {
-      isRefreshing: false,
-      modules: [],
-      dataSource,
-    }
+  state = {
+    isRefreshing: false,
+    rowData: [],
   }
 
   componentDidMount() {
-    const rows: DataSourceRow[] = []
-    const artworkModules = this.props.forYou.artwork_modules || []
-    const artistModules = this.props.forYou.artist_modules && this.props.forYou.artist_modules.concat()
-    const fairsModule = this.props.forYou.fairs_module
-    rows.push({ type: "fairs", data: fairsModule })
-    for (let i = 0; i < artworkModules.length; i++) {
-      const artworkModule = artworkModules[i]
-      rows.push({ type: "artwork", data: artworkModule })
-      if ((i + 1) % 2 === 0) {
+    const { forYou } = this.props
+    const rowData = []
+    const artworkModules = forYou.artwork_modules || []
+    const artistModules = forYou.artist_modules && forYou.artist_modules.concat()
+    const fairsModule = forYou.fairs_module
+
+    rowData.push({
+      type: "fairs",
+      data: fairsModule,
+    })
+
+    artworkModules.forEach((artworkModule, index) => {
+      rowData.push({
+        type: "artwork",
+        data: artworkModule,
+      })
+
+      const alternateRow = (index + 1) % 2 === 0
+
+      if (alternateRow) {
         const artistModule = artistModules.shift()
         if (artistModule) {
-          rows.push({ type: "artist", data: artistModule })
+          rowData.push({
+            type: "artist",
+            data: artistModule,
+          })
         }
       }
-    }
+    })
+
     this.setState({
-      dataSource: this.state.dataSource.cloneWithRows(rows),
+      rowData,
     })
   }
 
-  handleRefresh = () => {
-    this.setState({ isRefreshing: true })
-    const stopRefreshing = () => this.setState({ isRefreshing: false })
-
-    Promise.all(
-      this.state.modules.map(module => {
-        return new Promise((resolve, reject) => {
-          module.forceFetch(null, readyState => {
-            if (readyState.error) {
-              reject(readyState.error)
-            } else if (readyState.aborted) {
-              reject()
-            } else if (readyState.done) {
-              resolve()
-            }
+  handleRefresh = async () => {
+    this.setState({
+      isRefreshing: true,
+    })
+
+    try {
+      await Promise.all(
+        this.modules.map(module => {
+          return new Promise((resolve, reject) => {
+            const { props, state: { data, relayProp } } = module
```

Comment:
This doesn’t appear to follow the `Module` interface you’ve defined above.

I _think_ a bunch of this faffing with internals is because `ref` will return the intermediate Relay container, not our own actual component, right? I wonder if, instead of using `ref` it would make sense to have the rail components take a render callback with a reference to our own actual components and add methods to those rails that do their own force fetching.

```tsx
class ForYou extends React.Component {
  handleRefresh = async () => {
    await Promise.all(this.modules.map(module => module.refreshData()))
  }

  render() {
    return <ArtistRail  key={data.__id} rail={data}>{registerModule}</ArtistRail>
  }
}

class ArtistRail extends React.Component {
  refreshData() {
    return new Promise((resolve, reject) {
      this.props.relay.refetch(…)
      resolve()
    })
  }

  render() {
    const renderCallback = this.children
    renderCallback(this)
    return <View>…</View>
  }
}
```

Reply:
I also wonder if it’s possible for Relay, and if it would make sense, to pass the `ref` prop on to the wrapped component instead, but that’s a longer term thing to consider.

Reply:
The above def seems to encapsulate things better, let me give it a spin